napari/_vispy/layers/vectors.py

```diff
@@ -107,25 +112,231 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
+    Vertices are generated according to the following scheme:
+        1---x---0
+        | .     |
+        |   .   |
+        |     . |
+    5---3-------2---4
+       .         .
+          .   .
+            6
+
+    Where x marks the start point of the vector, and the vertex 6 its end
+    point.
+
+    In the case of k 2D vectors, the output 'triangles' is:
+    [
+        [0,1,2],                # vector 0,   triangle i=0
+        [1,2,3],                # vector 0,   triangle i=1
+        [4,5,6],                # vector 0,   triangle i=2
+        [7,8,9],                # vector 1,   triangle i=3
+        [8,9,10],               # vector 1,   triangle i=4
+        [11,12,13],             # vector 1,   triangle i=5
+
+        ...,
+
+        [7i/3,           7i/3 + 1,       7i/3 + 2],
+            # vector k-1, triangle i=3k-3 (i%3=0)
+        [7(i - 1)/3 + 1, 7(i - 1)/3 + 2, 7(i - 1)/3 + 3],
+            # vector k-1, triangle i=3k-2 (i%3=1)
+        [7(i - 2)/3 + 4, 7(i - 2)/3 + 5, 7(i - 2)/3 + 6]
+            # vector k-1, triangle i=3k-1 (i%3=2)
+    ]
+    """
+    nvectors, _, ndim = vectors.shape
+
+    vectors_starts = vectors[:, 0]
+
+    # Will be used to generate the vertices 2,3,4 and 5.
+    # Right now the head of the arrow is put at 75% of the length
+    # of the vector.
+    vectors_intermediates = vectors_starts + 0.75 * length * vectors[:, 1]
+
+    vectors_ends = vectors_starts + length * vectors[:, 1]
+
+    vertices = np.zeros((7 * nvectors, ndim))
+    offsets = segment_normal(vectors_starts, vectors_ends, p=p)
+    offsets = np.repeat(offsets, 7, axis=0)
+
+    signs = np.ones((len(offsets), ndim))
+    signs[::2] = -1
+    multipliers = np.ones((len(offsets), ndim))
+    multipliers[4::7] = 2
+    multipliers[5::7] = 2
+    multipliers[6::7] = 0
+
+    # here 'multipliers' is used to prevent vertex 6 from being offset,
+    # and to offset vertices 4 and 5 twice as much as vertices 2 and 3
+    offsets = offsets * signs * multipliers
+
+    vertices[::7] = vectors_starts
+    vertices[1::7] = vectors_starts
+    vertices[2::7] = vectors_intermediates
+    vertices[3::7] = vectors_intermediates
+    vertices[4::7] = vectors_intermediates
+    vertices[5::7] = vectors_intermediates
+    vertices[6::7] = vectors_ends
+
+    vertices = vertices + width * offsets / 2
+
+    triangles = np.array(
+        [
+            [7 * i / 3, 7 * i / 3 + 1, 7 * i / 3 + 2]
+            if i % 3 == 0
+            else [
+                7 * (i - 1) / 3 + 1,
+                7 * (i - 1) / 3 + 2,
+                7 * (i - 1) / 3 + 3,
+            ]
+            if i % 3 == 1
+            else [
+                7 * (i - 2) / 3 + 4,
+                7 * (i - 2) / 3 + 5,
+                7 * (i - 2) / 3 + 6,
+            ]
+            for i in range(3 * nvectors)
         ]
     ).astype(np.uint32)
 
+    # np.tile([[0,1,2],[1,2,3],[4,5,6]], (nvectors, 1)) + np.repeat(7*np.arange(nvectors), 3)[:,None]
```

Comment:
This didn't work? I think you're very close! 😃

Reply:
Thank you for your feedback and the time you took to dive into the details of the code !
I have a busy week but I will try to implement your suggestions by Friday.

For this suggestion in particular, I found that this one-liner was probably faster, but definitely really hard to interpret at first glance. Do you think it would be a better idea to reimplement such lines (instead of the very broken down list comprehensions that are currently implemented) and to comment them profusely instead ?

Reply:
Definitely favour the speed in this part of the codebase. If it helps, you could leave a pseudocode version of the comprehension in a comment to explain what's going on.

Reply:
I reverted back to my previous implementation and tried to comment the process. I'll resolve the suggestion if you feel like it is enough. :wink: